insights/parsers/__init__.py

```diff
@@ -498,38 +506,90 @@ def keyword_search(rows, **kwargs):
          {'domain': 'root', 'type': 'soft', 'item': 'nproc', 'value': -1}]
         >>> keyword_search(rows, domain__startswith='r')
         [{'domain': 'root', 'type': 'soft', 'item': 'nproc', 'value': -1}]
+
+    Testing has shown that caching the keyword_search() function itself does
+    not result in much speed-up, but caching the key transformation does.  The
+    cache is stored as an attribute, either on the object storing the rows or
+    on a 'parent' object that can take an attribute (if 'rows' is a list, that
+    cannot have an attribute added to it).  (We used to store the transformed
+    dictionary of rows, but storing just the key transformations is faster.)
     """
-    results = []
     if not kwargs:
-        return results
+        return []
+    if not rows:
+        return []
 
     # Allows us to transform the key and do lookups like __contains and
     # __startswith
     matchers = {
-        'default': lambda s, v: s == v,
+        'equals': lambda s, v: s == v,
         'contains': lambda s, v: s is not None and v in s,
         'startswith': lambda s, v: s is not None and s.startswith(v),
         'endswith': lambda s, v: s is not None and s.endswith(v),
         'lower_value': lambda s, v: None not in (s, v) and s.lower() == v.lower(),
     }
 
-    def key_match(row, key, value):
-        # Translate ' ' and '-' of keys in dict to '_' to match keyword arguments.
-        my_row = {}
-        for my_key, val in row.items():
-            my_row[my_key.replace(' ', '_').replace('-', '_')] = val
-        matcher_fn = matchers['default']
-        if '__' in key:
-            key, matcher = key.split('__', 1)
+    txform_cache_attr = '_transform_cache'
+    if parent is None and hasattr(rows, '__dict__'):
+        parent = rows
+    # Uncomment this 'if' to check that all the parsers are supplying an
+    # object, somehow, that can store our transformed row cache.
+    # if parent is None:
+    #     print("Invoked with no parent arg on primitive container - use parent=self argument to cache row transform")
+    # The actual txkeys cache gets used rarely - in a basic test with a real
+    # archive it only got used eleven times.  But that's a saving...
+    if parent is not None and hasattr(parent, txform_cache_attr):
+        txkeys = getattr(parent, txform_cache_attr)
+    else:
+        # Store the translation from the search key to the key in the data.
+        # We start with the reverse transform - the key, to the transformed
+        # key.  This avoids doing the string replacement for every key in
+        # every row.  Then we invert that to map from the search term (with
+        # spaces and dashes replaced by underscores) to the key in the row
+        # data.
+        rev_txkeys = dict()
+        for row in rows:
```

Comment:
Is it necessary to iterator all `rows`? Generally, one line is okay, as each row should have the same `keys`. Imagine the content is huge, this `for` loop would be time-consuming.
Or, if you are worried about different keys in rows, how about adding one local cache to fast it?:

```python
_raw_keys = set()
for row in rows:
    if _raw_keys.issuperset(row.keys()):
        continue
    _raw_keys.update(row.keys())
for key in _raw_keys:
    if key not in rev_txkeys:
        rev_txkeys[key] = key.replace(' ', '_').replace('-', '_')
```